File: java/src/main/java/serie2/ConcurrentQueue.java
package serie2;

import java.util.concurrent.atomic.AtomicReference;

public class ConcurrentQueue<T> {
    private static class Node<E>{
        final E item;
        final AtomicReference<Node<E>> next;
        public Node(E item, Node<E> next){
            this.item = item;
            this.next = new AtomicReference<>(next);
        }
    }
    private final AtomicReference<Node<T>> head;
    private final AtomicReference<Node<T>> tail;

    public ConcurrentQueue(){
        Node<T> dummy = new Node<T>(null,null);
        this.head = new AtomicReference<>(dummy);
        this.tail = new AtomicReference<>(dummy);
    }

    public void put(T item){
        Node<T> newNode = new Node<T>(item, null);
        while (true) {
            Node<T> currTail = tail.get();
            Node<T> tailNext = currTail.next.get();
            if (tail.get()==currTail) { // Are tail and next consistent?
                if (tailNext != null) {
                    // queue in intermediate state, advance state
                    tail.compareAndSet(currTail, tailNext);
                } else {
                    if (currTail.next.compareAndSet(null, newNode)) {
                        // insertion succeeded, try advancing tail
                        tail.compareAndSet(currTail, newNode);
                        return;
                    }
                }
            }
        }
    }

    public T tryTake() {
        while(true) {
            Node<T> headCur = head.get();
            Node<T> tailCur = tail.get();
            Node<T> headNext = headCur.next.get();

            if(headCur == head.get()) {
                if(headCur == tailCur) {
                    if(headNext == null)
                        return null;

                    tail.compareAndSet(tailCur, headNext);
                } else   {
                    T val = headNext.item;
                    if(head.compareAndSet(headCur, headNext))
                        return val;
                }
            }
        }
    }

    // take a datum - spinning if necessary
    public T take() throws InterruptedException {
        T v;
        while ((v = tryTake()) == null) {
            Thread.sleep(0);
        }
        return v;
    }

    public boolean isEmpty() {
        return tail.get() == head.get();
    }

}

File: pc1718i/ConcurrencyProgramming/serie2/ConcurrentQueue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace ConcurrencyProgramming.serie2
{
    class ConcurrentQueue <T>{
		private class Node<E>{
            public readonly E Item;
            public volatile Node<E> Next;
            public Node(E item,Node<E> next){
                this.Item = item;
                this.Next = next;
            }
		}

        private volatile Node<T> head;
        private volatile Node<T> tail;

        public ConcurrentQueue(){
            Node<T> dummy = new Node<T>(default(T), null);
            this.head = dummy;
            this.tail = dummy;
        }

        public void Put(T item){
            Node<T> newNode = new Node<T>(item, null);
            Node<T> currTail;
            Node<T> tailNext;
            while(true){
                currTail = tail;
                tailNext = currTail.Next;
                if(currTail == tail){
                    if(tailNext != null){
                         Interlocked.CompareExchange(ref tail, tailNext, currTail);
                    }
                    else{
                        if(Interlocked.CompareExchange(ref currTail.Next,newNode,null) == null){
                            Interlocked.CompareExchange(ref tail, newNode, currTail);
                            return;
                        }
                    }
                }
            }
        }

        public T TryTake(){
            Node<T> headCurr;
            Node<T> tailCurr;
            Node<T> headNext;
            while(true){
                headCurr = head;
                tailCurr = tail;
                headNext = headCurr.Next;
                if(headCurr == head){
                    if(headCurr == tailCurr){
                        if(headNext == null){
                            return default(T);
                        }
                        Interlocked.CompareExchange(ref tail, headNext, tailCurr);
                    }
                    else{
                        if(Interlocked.CompareExchange(ref head, headNext, headCurr) == headCurr){
                            return headNext.Item;
                        }
                    }
                }
            }
        }

        public T Take(){
            T v;
            while((v = TryTake()) == null){
                Thread.Sleep(0);
            }
            return v;
        }

        public bool IsEmpty(){
            Node<T> currTail = tail;
            return Interlocked.CompareExchange(ref head, currTail, currTail) == currTail;
        }

	}
    

}
